emu/src/cpu/arm7tdmi.rs

```diff
@@ -770,6 +772,42 @@ impl Arm7tdmi {
             _ => unreachable!(),
         }
     }
+
+    pub(crate) fn push_pop_register(&mut self, op_code: ThumbModeOpcode) -> Option<u32> {
+        let load_store: LoadStoreKind = op_code.get_bit(11).into();
+        let store_or_not_reg_lr = op_code.get_bit(8);
+        let rlist = op_code.get_bits(0..=7);
+
+        match load_store {
+            LoadStoreKind::Store if !store_or_not_reg_lr => todo!(),
+            LoadStoreKind::Store if store_or_not_reg_lr => {
+                let reg_sp = self.registers.register_at(REG_SP);
+                self.registers.set_register_at(REG_SP, reg_sp - 4);
+                let stack_addr = (self.registers.register_at(REG_SP) & !3) as usize;
+                self.memory
+                    .lock()
+                    .unwrap()
+                    .write_word(stack_addr, self.registers.register_at(REG_LR));
+
+                for r in (0..8).rev() {
+                    if rlist.get_bit(r) {
+                        let reg_sp = self.registers.register_at(REG_SP);
+                        self.registers.set_register_at(REG_SP, reg_sp - 4);
+                        let stack_addr = (self.registers.register_at(REG_SP) & !3) as usize;
+                        self.memory
+                            .lock()
+                            .unwrap()
+                            .write_word(stack_addr, self.registers.register_at(r.into()));
+                    }
+                }
```

Comment:
LGTM in general, I would only rewrite it in a way where we don't read/store every time the sp but we read it once at the beginning, we use it, we decrement it each time and then only at the end we write it back.

Also, I'm not sure about the `& !3`, I know it's in order to have a word-aligned address but documentation says that the 2 lsb might be interpreted by the memory system (maybe we will need to do it in the future?). And also in the implementation of the ldm/stm we are not doing `& !3`. Maybe we can avoid doing it and we put a warning in the `write_word`/`read_word` in the `InternalMemory` if the provided address is not word-aligned? So we know if something is off or if we should interpret them somehow

Reply:
Right now I removed! I will add a little warning in r/w word